test/e2e/validation.go

```diff
@@ -22,199 +22,211 @@ import (
-		// webhook works across all namespaces.
-		validKFC.Namespace = f.TestNamespaceName()
-	})
-
-	It(fmt.Sprintf("should fail when an invalid %s is created or updated", resourceName), func() {
-		// This test also implicitly tests the successful creation of a valid
-		// resource.
-		By(fmt.Sprintf("Creating an invalid %s", resourceName))
-		invalidKFC := validKFC.DeepCopyObject().(*v1beta1.KubeFedCluster)
-		invalidKFC.Spec.APIEndpoint = ""
-		err := client.Create(context.TODO(), invalidKFC)
-		if err == nil {
-			f.Logger().Fatalf("Expected error creating invalid %s = %+v", resourceName, invalidKFC)
-		}
-
-		By(fmt.Sprintf("Creating a valid %s", resourceName))
-		validKFCCopy := validKFC.DeepCopyObject().(*v1beta1.KubeFedCluster)
-		err = client.Create(context.TODO(), validKFCCopy)
-		if err != nil {
-			f.Logger().Fatalf("Unexpected error creating valid %s = %+v, err: %v", resourceName, validKFCCopy, err)
-		}
-
-		By(fmt.Sprintf("Updating with an invalid %s", resourceName))
-		invalidKFC = validKFCCopy.DeepCopyObject().(*v1beta1.KubeFedCluster)
-		invalidKFC.Spec.SecretRef.Name = ""
-		err = client.Update(context.TODO(), invalidKFC)
-		if err == nil {
-			f.Logger().Fatalf("Expected error updating invalid %s = %+v", resourceName, invalidKFC)
-		}
-
-		err = client.Delete(context.TODO(), validKFCCopy, validKFCCopy.Namespace, validKFCCopy.Name)
-		if err != nil && !apierrors.IsNotFound(err) {
-			f.Logger().Errorf("Error deleting %s %s: %v", resourceName, validKFCCopy.Name, err)
-		}
-	})
-
-	When("running with namespace scoped deployment", func() {
-		It(fmt.Sprintf("should succeed when an invalid %s is created outside the kubefed system namespace", resourceName), func() {
-			if !framework.TestContext.LimitedScope {
-				framework.Skipf("Cannot run validation admission webhook namespaced test in a cluster scoped deployment")
+			By(fmt.Sprintf("Updating with an invalid %s", resourceName))
+			invalidObj = vrt.invalidObjectFromValid(validObj)
+			err = client.Update(context.TODO(), invalidObj)
+			if err == nil {
+				f.Logger().Fatalf("Expected error updating invalid %s = %+v", resourceName, vrt)
 			}
-			kubeClient := f.KubeClient(fmt.Sprintf("%s-create-namespace", testBaseName))
-			invalidKFC := validKFC.DeepCopyObject().(*v1beta1.KubeFedCluster)
-			namespace := framework.CreateTestNamespace(kubeClient, testBaseName)
-			framework.AddCleanupAction(func() {
-				framework.DeleteNamespace(kubeClient, namespace)
-			})
 
-			By(fmt.Sprintf("Creating an invalid %s in the separate test namespace %s", resourceName, namespace))
-			invalidKFC.Namespace = namespace
-			invalidKFC.Spec.APIEndpoint = "https://"
-			err := client.Create(context.TODO(), invalidKFC)
-			if err != nil {
-				f.Logger().Fatalf("Unexpected error creating invalid %s = %+v in another test namespace %s, err: %v", resourceName, invalidKFC, namespace, err)
+			// Immediately delete the created test resource to avoid errors in
```

Comment:
I'm not sure this is a sufficient guarantee of test isolation.  How about running the tests only for in-process controllers in cluster-scoped mode to allow for resource creation to be isolated to a test namespace? The webhook would still be running. A check for the webhook being correctly configured for a namespaced control plane could consist of dry-run attempts to create valid and invalid resources to avoid impacting controllers targeting the system namespace.

Reply:
I think I understand the concern, but I'm not sure I understand the proposed solution alternatives.

For the concern: since we run tests serially, there's probably not too much harm in creating temporary resources that could impact the control plane only during the execution of this test, since the resources are immediately deleted before proceeding. Are you thinking about parallelization?

For the proposed solutions: cluster-scoped mode will run in a separate test namespace regardless of in-process controllers. So I think we're okay there. When running against a namespaced control plane, we could attempt to create a separate test namespace just for this test once a singleton webhook becomes a thing. Until then, how would we perform dry-run attempts to verify validation of an update of already created resources? Or are you thinking to forego the update validation check?

Reply:
I am not concerned with running tests in parallel.  I think it would be preferable to avoid creating resources that will impact controller state. As you say, this isn't a problem for a cluster-scoped control plane since a test namespace can be used. For a namespace-scoped control plane, I think a preferable strategy to the current approach would be to use dry-run validation. In the case of an update validation check, it should be possible to target the FTC of a type enabled by default (e.g. `deployments.app`).

Reply:
I think that's doable and a good approach for a future follow-up.